fix: respect package.json type field in plugin-resolved modules

hyf0 opened the pull request:
partially fixes, #6398

When plugins resolve module IDs, ModuleDefFormat now checks both file extensions and package.json "type" field, matching the behavior of the internal resolver. This ensures consistent module format determination regardless of resolution path.

Changes:
- Add infer_module_def_format() helper in resolve_id_with_plugins.rs
- Update resolve_dynamic_import and resolve_id hooks to use helper
- Add test case with ESM package resolved via internal resolver

hyf0 commented:
* **#6400** <a href="https://app.graphite.dev/github/pr/rolldown/rolldown/6400?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/rolldown/rolldown/6400?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add the label <kbd>graphite: merge</kbd> to this PR to add it to the merge queue.

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/rolldown?ref=merge-queue-instructions-in-stack-comment&prId=6400).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

sapphi-red reviewed (approved):
This would fix for plugins that uses `this.resolve`, but this would not fix the case for rolldown-vite because `resolver.package_json_cache()` is populated only when the internal resolver is used.

hyf0 committed:
fix: respect package.json type field in plugin-resolved modules (#6400)

partially fixes, #6398

When plugins resolve module IDs, ModuleDefFormat now checks both file extensions and package.json "type" field, matching the behavior of the internal resolver. This ensures consistent module format determination regardless of resolution path.

Changes:
- Add infer_module_def_format() helper in resolve_id_with_plugins.rs
- Update resolve_dynamic_import and resolve_id hooks to use helper
- Add test case with ESM package resolved via internal resolver
